presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestShardRecovery.java

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.storage;
+
+import com.facebook.presto.RowPagesBuilder;
+import com.facebook.presto.metadata.InMemoryNodeManager;
+import com.facebook.presto.raptor.metadata.DatabaseShardManager;
+import com.facebook.presto.raptor.metadata.ShardManager;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import io.airlift.units.Duration;
+import org.skife.jdbi.v2.DBI;
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.IDBI;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.io.Files.createTempDir;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestShardRecovery
+{
+    private StorageService storageService;
+    private ShardRecoveryManager recoveryManager;
+    private Handle dummyHandle;
+
+    @BeforeMethod
+    public void setup()
+            throws Exception
+    {
+        File temporary = createTempDir();
+        File directory = new File(temporary, "data");
+        File backupDirectory = new File(temporary, "backup");
+        storageService = new FileStorageService(directory, Optional.of(backupDirectory));
+        storageService.start();
+
+        IDBI dbi = new DBI("jdbc:h2:mem:test" + System.nanoTime());
+        dummyHandle = dbi.open();
+        ShardManager shardManager = new DatabaseShardManager(dbi);
+        recoveryManager = new ShardRecoveryManager(storageService, new InMemoryNodeManager(), shardManager, new Duration(5, TimeUnit.MINUTES), 10);
+    }
+
+    @AfterMethod
+    public void tearDown()
+            throws Exception
+    {
+        dummyHandle.close();
+    }
+
+    @SuppressWarnings("EmptyTryBlock")
+    @Test
+    public void testShardRecovery()
+            throws Exception
+    {
+        List<Long> columnIds = ImmutableList.of(3L, 7L);
+        List<Type> columnTypes = ImmutableList.<Type>of(BIGINT, VARCHAR);
+
+        UUID shardUuid = UUID.randomUUID();
+        File file = storageService.getStorageFile(shardUuid);
+        File backupFile = storageService.getBackupFile(shardUuid);
+
+        try (OrcFileWriter writer = new OrcFileWriter(columnIds, columnTypes, backupFile)) {
+            // create empty file in storage
```

Comment:
This should say "create file with zero rows" (since the file won't technically be "empty")